File: sdk-go/session.go
package boxkite

import (
	"context"

	"github.com/gorilla/websocket"
)

// Session is a thin, session-scoped view over Client -- every method
// forwards to the identically-named Client method with this Session's ID
// as the sessionID argument. Obtained via WithSandbox; mirrors
// sdk-python's SandboxSession / sdk-js's SandboxSession.
type Session struct {
	client *Client
	ID     string
}

// WithSandbox is the create-on-enter, destroy-on-exit convenience --
// mirroring sdk-python's `with client.sandbox() as sb:` context manager /
// sdk-js's withSandbox. It creates a sandbox from req, invokes fn with a
// *Session bound to it, and destroys the sandbox when fn returns (even if
// fn returns an error) -- teardown itself is best-effort, mirroring both
// sibling SDKs' "an already-gone session shouldn't raise on cleanup"
// behavior.
func (c *Client) WithSandbox(ctx context.Context, req CreateSandboxRequest, fn func(*Session) error) error {
	sandbox, err := c.CreateSandbox(ctx, req)
	if err != nil {
		return err
	}
	sess := &Session{client: c, ID: sandbox.ID}
	defer func() {
		_ = c.DestroySandbox(ctx, sandbox.ID)
	}()
	return fn(sess)
}

func (s *Session) Exec(ctx context.Context, command string, opts *ExecOptions) (*ExecResult, error) {
	return s.client.Exec(ctx, s.ID, command, opts)
}

func (s *Session) HTTPRequest(ctx context.Context, method, targetURL string, opts *HTTPRequestOptions) (*HTTPRequestResult, error) {
	return s.client.HTTPRequest(ctx, s.ID, method, targetURL, opts)
}

func (s *Session) FileCreate(ctx context.Context, path, content string, opts *FileOptions) (*FileCreateResult, error) {
	return s.client.FileCreate(ctx, s.ID, path, content, opts)
}

func (s *Session) View(ctx context.Context, path string, opts *ViewOptions) (*FileViewResult, error) {
	return s.client.View(ctx, s.ID, path, opts)
}

func (s *Session) StrReplace(ctx context.Context, path, oldStr, newStr string, opts *StrReplaceOptions) (*StrReplaceResult, error) {
	return s.client.StrReplace(ctx, s.ID, path, oldStr, newStr, opts)
}

func (s *Session) Ls(ctx context.Context, path string) (*LsResult, error) {
	return s.client.Ls(ctx, s.ID, path)
}

func (s *Session) Glob(ctx context.Context, pattern, path string) (*GlobResult, error) {
	return s.client.Glob(ctx, s.ID, pattern, path)
}

func (s *Session) Grep(ctx context.Context, pattern string, opts *GrepOptions) (*GrepResult, error) {
	return s.client.Grep(ctx, s.ID, pattern, opts)
}

func (s *Session) GetLog(ctx context.Context, opts *GetLogOptions) (*GetLogResult, error) {
	return s.client.GetLog(ctx, s.ID, opts)
}

func (s *Session) Watch(ctx context.Context) (*LogWatcher, error) {
	return s.client.Watch(ctx, s.ID)
}

func (s *Session) StartProcess(ctx context.Context, command string, opts *StartProcessOptions) (*ProcessStartResult, error) {
	return s.client.StartProcess(ctx, s.ID, command, opts)
}

func (s *Session) ListProcesses(ctx context.Context) (*ProcessListResult, error) {
	return s.client.ListProcesses(ctx, s.ID)
}

func (s *Session) GetProcessOutput(ctx context.Context, processID string, sinceOffset int64) (*ProcessOutputResult, error) {
	return s.client.GetProcessOutput(ctx, s.ID, processID, sinceOffset)
}

func (s *Session) SendProcessInput(ctx context.Context, processID, data string) (*ProcessInputResult, error) {
	return s.client.SendProcessInput(ctx, s.ID, processID, data)
}

func (s *Session) StopProcess(ctx context.Context, processID string) (*ProcessStopResult, error) {
	return s.client.StopProcess(ctx, s.ID, processID)
}

func (s *Session) Takeover(ctx context.Context) (*websocket.Conn, error) {
	return s.client.Takeover(ctx, s.ID)
}

func (s *Session) DesktopTakeover(ctx context.Context) (*websocket.Conn, error) {
	return s.client.DesktopTakeover(ctx, s.ID)
}

func (s *Session) CreatePreviewURL(ctx context.Context, port int, ttlSeconds *int) (*PreviewURL, error) {
	return s.client.CreatePreviewURL(ctx, s.ID, port, ttlSeconds)
}

func (s *Session) RevokePreviewURL(ctx context.Context, port int, tokenID string) (*PreviewRevokeResult, error) {
	return s.client.RevokePreviewURL(ctx, s.ID, port, tokenID)
}

File: sdk-rust/src/session.rs
//! [`SandboxSession`]: a thin, session-scoped view over [`Client`] whose
//! every method forwards to the identically-named `Client` method with this
//! session's id bound in, plus [`Client::with_sandbox`], the
//! create-on-enter/destroy-on-exit convenience. Mirrors `sdk-python`'s
//! `SandboxSession`/`client.sandbox()` context manager, `sdk-js`'s
//! `SandboxSession`/`withSandbox`, and `sdk-go`'s `Session`/`WithSandbox`.

use std::future::Future;
use std::pin::Pin;

use futures_util::Stream;

use crate::audit::{AuditLogEntry, AuditLogResponse, GetLogOptions};
use crate::client::Client;
use crate::desktop::DesktopStream;
use crate::error::BoxkiteError;
use crate::files::{
    ExecOptions, ExecResult, FileCreateResult, FileOptions, GlobOptions, GlobResult, GrepOptions,
    GrepResult, HttpRequestOptions, HttpRequestResult, LsOptions, LsResult, StrReplaceOptions,
    StrReplaceResult, ViewOptions, ViewResult,
};
use crate::lsp::{LspCompletionResult, LspOpenResult, LspStartResult, LspStopResult};
use crate::preview::{PreviewRevokeResult, PreviewUrl};
use crate::processes::{
    ProcessInputResult, ProcessListResult, ProcessOutputResult, ProcessStartResult,
    ProcessStopResult, StartProcessOptions,
};
use crate::sandboxes::CreateSandboxOptions;
use crate::takeover::TakeoverStream;

/// A [`Client`] bound to one sandbox session id -- obtained from
/// [`Client::with_sandbox`]. Holds a cheap clone of the underlying `Client`
/// (see [`Client`]'s own note on cloning), so it is `Clone` and can outlive
/// the borrow that produced it.
#[derive(Clone, Debug)]
pub struct SandboxSession {
    client: Client,
    id: String,
}

impl SandboxSession {
    /// The bound sandbox session id.
    pub fn id(&self) -> &str {
        &self.id
    }

    /// The underlying [`Client`], for the handful of calls not scoped to a
    /// single session (e.g. [`Client::list_sandboxes`]).
    pub fn client(&self) -> &Client {
        &self.client
    }

    pub async fn exec(
        &self,
        command: &str,
        options: ExecOptions,
    ) -> Result<ExecResult, BoxkiteError> {
        self.client.exec(&self.id, command, options).await
    }

    pub async fn http_request(
        &self,
        method: &str,
        url: &str,
        options: HttpRequestOptions,
    ) -> Result<HttpRequestResult, BoxkiteError> {
        self.client
            .http_request(&self.id, method, url, options)
            .await
    }

    pub async fn file_create(
        &self,
        path: &str,
        content: &str,
        options: FileOptions,
    ) -> Result<FileCreateResult, BoxkiteError> {
        self.client
            .file_create(&self.id, path, content, options)
            .await
    }

    pub async fn view(&self, path: &str, options: ViewOptions) -> Result<ViewResult, BoxkiteError> {
        self.client.view(&self.id, path, options).await
    }

    pub async fn str_replace(
        &self,
        path: &str,
        old_str: &str,
        new_str: &str,
        options: StrReplaceOptions,
    ) -> Result<StrReplaceResult, BoxkiteError> {
        self.client
            .str_replace(&self.id, path, old_str, new_str, options)
            .await
    }

    pub async fn ls(&self, options: LsOptions) -> Result<LsResult, BoxkiteError> {
        self.client.ls(&self.id, options).await
    }

    pub async fn glob(
        &self,
        pattern: &str,
        options: GlobOptions,
    ) -> Result<GlobResult, BoxkiteError> {
        self.client.glob(&self.id, pattern, options).await
    }

    pub async fn grep(
        &self,
        pattern: &str,
        options: GrepOptions,
    ) -> Result<GrepResult, BoxkiteError> {
        self.client.grep(&self.id, pattern, options).await
    }

    pub async fn get_log(&self, options: GetLogOptions) -> Result<AuditLogResponse, BoxkiteError> {
        self.client.get_log(&self.id, options).await
    }

    pub fn watch(
        &self,
    ) -> Pin<Box<dyn Stream<Item = Result<AuditLogEntry, BoxkiteError>> + Send + 'static>> {
        self.client.watch(&self.id)
    }

    pub async fn start_process(
        &self,
        command: &str,
        options: StartProcessOptions,
    ) -> Result<ProcessStartResult, BoxkiteError> {
        self.client.start_process(&self.id, command, options).await
    }

    pub async fn list_processes(&self) -> Result<ProcessListResult, BoxkiteError> {
        self.client.list_processes(&self.id).await
    }

    pub async fn get_process_output(
        &self,
        process_id: &str,
        since_offset: i64,
    ) -> Result<ProcessOutputResult, BoxkiteError> {
        self.client
            .get_process_output(&self.id, process_id, since_offset)
            .await
    }

    pub async fn send_process_input(
        &self,
        process_id: &str,
        data: &str,
    ) -> Result<ProcessInputResult, BoxkiteError> {
        self.client
            .send_process_input(&self.id, process_id, data)
            .await
    }

    pub async fn stop_process(&self, process_id: &str) -> Result<ProcessStopResult, BoxkiteError> {
        self.client.stop_process(&self.id, process_id).await
    }

    pub async fn takeover(&self) -> Result<TakeoverStream, BoxkiteError> {
        self.client.takeover(&self.id).await
    }

    pub async fn desktop_takeover(&self) -> Result<DesktopStream, BoxkiteError> {
        self.client.desktop_takeover(&self.id).await
    }

    pub async fn create_preview_url(
        &self,
        port: u16,
        ttl_seconds: Option<u32>,
    ) -> Result<PreviewUrl, BoxkiteError> {
        self.client
            .create_preview_url(&self.id, port, ttl_seconds)
            .await
    }

    pub async fn revoke_preview_url(
        &self,
        port: u16,
        token_id: &str,
    ) -> Result<PreviewRevokeResult, BoxkiteError> {
        self.client
            .revoke_preview_url(&self.id, port, token_id)
            .await
    }

    pub async fn lsp_start(&self, language: &str) -> Result<LspStartResult, BoxkiteError> {
        self.client.lsp_start(&self.id, language).await
    }

    pub async fn lsp_open(
        &self,
        lsp_id: &str,
        path: &str,
        content: &str,
    ) -> Result<LspOpenResult, BoxkiteError> {
        self.client.lsp_open(&self.id, lsp_id, path, content).await
    }

    pub async fn lsp_completion(
        &self,
        lsp_id: &str,
        path: &str,
        line: u32,
        character: u32,
    ) -> Result<LspCompletionResult, BoxkiteError> {
        self.client
            .lsp_completion(&self.id, lsp_id, path, line, character)
            .await
    }

    pub async fn lsp_stop(&self, lsp_id: &str) -> Result<LspStopResult, BoxkiteError> {
        self.client.lsp_stop(&self.id, lsp_id).await
    }
}

impl Client {
    /// Create a sandbox, run `f` against a [`SandboxSession`] bound to it, and
    /// tear the sandbox down when `f` returns -- even on error. Teardown is
    /// best-effort (an already-gone session doesn't turn a successful `f` into
    /// an error), mirroring every sibling SDK's cleanup semantics.
    ///
    /// ```no_run
    /// use boxkite_client::{Client, CreateSandboxOptions, ExecOptions};
    ///
    /// # async fn example() -> Result<(), boxkite_client::BoxkiteError> {
    /// let client = Client::new("https://cp.example.com", "bxk_live_...")?;
    /// let stdout = client
    ///     .with_sandbox(CreateSandboxOptions::new().label("demo"), |sb| async move {
    ///         let result = sb.exec("echo hi", ExecOptions::new()).await?;
    ///         Ok(result.stdout)
    ///     })
    ///     .await?;
    /// assert_eq!(stdout, "hi\n");
    /// # Ok(())
    /// # }
    /// ```
    pub async fn with_sandbox<F, Fut, T>(
        &self,
        options: CreateSandboxOptions,
        f: F,
    ) -> Result<T, BoxkiteError>
    where
        F: FnOnce(SandboxSession) -> Fut,
        Fut: Future<Output = Result<T, BoxkiteError>>,
    {
        let sandbox = self.create_sandbox(options).await?;
        let session = SandboxSession {
            client: self.clone(),
            id: sandbox.id.clone(),
        };
        let result = f(session).await;
        // Best-effort teardown: don't let a cleanup failure mask f's result.
        let _ = self.destroy_sandbox(&sandbox.id).await;
        result
    }
}
